### PR TITLE
ref(weekly report): Prevent resending emails

### LINKED ISSUE #55476: Prevent resending weekly report emails
We recently had an error in the weekly report task. Since it has retries enabled it would get halfway through sending them and re-attempt the entire org.

Some possible solutions:
- Build the user context as a step and send them in a followup step so no emails are sent until all the contexts have been generated
- Limit the task to only retry on specific errors

ACs:
- Add a metric alert for the weekly_report task
- Limit retries to only if no emails have been sent

### PR DESCRIPTION
**Pull request by ceorourke**
Build up the user context per organization first, then send emails. 

Closes https://github.com/getsentry/sentry/issues/55476

### CODE REVIEW

File: src/sentry/tasks/weekly_reports.py [hunk 1]
```diff
@@ -547,17 +551,23 @@ def fetch_key_performance_issue_groups(ctx):
...
+    """
     # Specify a sentry user to send this email.
     if email_override:
         target_user_id = (
             target_user.id if target_user else None
         )  # if None, generates report for a user with access to all projects
-        send_email(ctx, target_user_id, dry_run=dry_run, email_override=email_override)
+        template_by_user_id = prepare_template_context(ctx, [target_user_id])
+        send_email(
+            ctx,
```
Comment (nhsiehgit) [inline src/sentry/tasks/weekly_reports.py:565]: can we make all of these kwargs?

File: src/sentry/tasks/weekly_reports.py [hunk 2]
```diff
@@ -945,16 +960,21 @@ def issue_summary():
...
-def send_email(ctx, user_id, dry_run=False, email_override=None):
-    template_ctx = render_template_context(ctx, user_id)
-    if not template_ctx:
-        logger.debug(
-            "Skipping report for %s to <User: %s>, no qualifying reports to deliver.",
-            ctx.organization.id,
-            user_id,
-        )
-        return
+def prepare_template_context(ctx, user_ids):
```
Comment (nhsiehgit) [inline src/sentry/tasks/weekly_reports.py:963]: mind adding typing here?
both to the arguments and return typing

File: src/sentry/tasks/weekly_reports.py [hunk 3]
```diff
@@ -571,8 +581,13 @@ def deliver_reports(ctx, dry_run=False, target_user=None, email_override=None):
         user_ids = notifications_service.get_users_for_weekly_reports(
             organization_id=ctx.organization.id, user_ids=user_list
         )
+        template_by_user_id = []
         for user_id in user_ids:
-            send_email(ctx, user_id, dry_run=dry_run)
+            template_by_user_id = prepare_template_context(ctx, user_ids)
+        for user_template in template_by_user_id:
+            send_email(
```
Comment (nhsiehgit) [inline src/sentry/tasks/weekly_reports.py:588]: default to using kwargs for clarity

File: src/sentry/tasks/weekly_reports.py [hunk 4]
```diff
@@ -547,17 +551,23 @@ def fetch_key_performance_issue_groups(ctx):
...
+    """
+    For all users in the organization, we generate the template context for the user, and send the email.
+    """
     # Specify a sentry user to send this email.
     if email_override:
         target_user_id = (
             target_user.id if target_user else None
         )  # if None, generates report for a user with access to all projects
-        send_email(ctx, target_user_id, dry_run=dry_run, email_override=email_override)
+        template_by_user_id = prepare_template_context(ctx, [target_user_id])
```
Comment (nhsiehgit) [inline src/sentry/tasks/weekly_reports.py:563]: sounds like this is a list of templates?
can we rename this to better represent the return value? Maybe something like `user_template_context_list`

**Review by nhsiehgit — APPROVED** [review]
nice.

Looks like a solid change but i don't have much context on the issue 😬 

a few nit comments, but non -blocking

File: src/sentry/tasks/weekly_reports.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/tasks/weekly_reports.py:563]: Yeah it's a list of dicts like `[{"context": ..., "user_id": 1}]`

**Review by 17hogeju — APPROVED** [review]
lgtm

**Pull request merged**